### PR TITLE
migrate to nom 5

### PR DESCRIPTION
**Pull request by NamsooCho**

### CODE REVIEW

**Comment by kpp** [comment]
Did you fix nom5?

**Comment by NamsooCho** [comment]
>Did you fix nom5?

I did nothing about nom5

File: Cargo.toml [hunk 1]
```diff
@@ -34,7 +34,7 @@ byteorder = "1"
 futures = "0.1"
 log = "0.4"
 sodiumoxide = "0.2.2"
-nom = "3.2"
+nom = { git = "https://github.com/Geal/nom.git" }
```
Comment (kurnevsky) [inline Cargo.toml]: There is `5.0.0-beta2` release of nom.

File: src/toxcore/dht/codec.rs [hunk 2]
```diff
@@ -132,9 +134,13 @@ impl Decoder for DhtCodec {
         }
 
         match Packet::from_bytes(buf) {
-            IResult::Incomplete(needed) => Err(DecodeError::incomplete_packet(needed, buf.to_vec())),
-            IResult::Error(error) => Err(DecodeError::deserialize(error, buf.to_vec())),
-            IResult::Done(_, packet) => {
+            Err(Err::Incomplete(needed)) => Err(DecodeError::incomplete_packet(needed, buf.to_vec())),
```
Comment (kurnevsky) [inline src/toxcore/dht/codec.rs]: Do we really need to handle all these variants? Why not just `Err(e) => ...`?

File: src/toxcore/dht/codec.rs [hunk 3]
```diff
@@ -132,9 +134,13 @@ impl Decoder for DhtCodec {
...
+            Err(Err::Error(error)) => {
+                let (_, kind) = error;
+                Err(DecodeError::deserialize(kind, buf.to_vec()))
+            },
+            Err(Err::Failure(e)) => panic!("Packet deserialize failed with unrecoverable error: {:?}", e),
```
Comment (kurnevsky) [inline src/toxcore/dht/codec.rs]: There mustn't be panics on packets decoding, otherwise it'll be easy to kill the node.

File: src/toxcore/dht/daemon_state.rs [hunk 4]
```diff
@@ -86,11 +88,17 @@ impl DaemonState {
     /// Deserialize DHT close list and then re-setup close list, old means that the format of deserialization is old version
     pub fn deserialize_old(server: &Server, serialized_data: &[u8]) -> impl Future<Item=(), Error=DeserializeError> + Send {
         let nodes = match DhtState::from_bytes(serialized_data) {
-            IResult::Done(_, DhtState(nodes)) => nodes,
-            IResult::Incomplete(needed) =>
-                return Either::A(future::err(DeserializeError::incomplete(needed, serialized_data.to_vec()))),
-            IResult::Error(error) =>
-                return Either::A(future::err(DeserializeError::deserialize(error, serialized_data.to_vec()))),
+            Err(Err::Incomplete(needed)) => {
```
Comment (kurnevsky) [inline src/toxcore/dht/daemon_state.rs]: I'd suggest to make one unified error type and don't handle all these variants.

File: src/toxcore/dht/daemon_state.rs [hunk 5]
```diff
@@ -148,7 +156,7 @@ mod tests {
         let serialized_len = serialized_vec.len();
         let res = DaemonState::deserialize_old(&alice, &serialized_vec[..serialized_len - 1]).wait();
         assert!(res.is_err());
-        assert_eq!(*res.err().unwrap().kind(), DeserializeErrorKind::IncompleteData { needed: Needed::Size(55), data: serialized_vec[..serialized_len - 1].to_vec() });
+        assert_eq!(*res.err().unwrap().kind(), DeserializeErrorKind::IncompleteData { needed: Needed::Size(39), data: serialized_vec[..serialized_len - 1].to_vec() });
```
Comment (kurnevsky) [inline src/toxcore/dht/daemon_state.rs]: Why has it changed?

File: src/toxcore/dht/packet/cookie.rs [hunk 6]
```diff
@@ -152,15 +154,18 @@ impl EncryptedCookie {
                 GetPayloadError::decrypt()
             })?;
         match Cookie::from_bytes(&decrypted) {
-            IResult::Incomplete(needed) => {
-                debug!(target: "Dht", "Cookie return deserialize error: {:?}", needed);
+            Err(Err::Incomplete(Needed::Unknown)) => {
```
Comment (kurnevsky) [inline src/toxcore/dht/packet/cookie.rs]: I'd suggest to make one unified error type and don't handle all these variants.

File: src/toxcore/onion/packet/mod.rs [hunk 7]
```diff
@@ -135,17 +138,20 @@ impl OnionReturn {
                 Error::new(ErrorKind::Other, "OnionReturn decrypt error.")
             })?;
         match OnionReturn::inner_from_bytes(&decrypted) {
-            IResult::Incomplete(e) => {
-                debug!(target: "Onion", "Inner onion return deserialize error: {:?}", e);
+            Err(Err::Incomplete(Needed::Unknown)) => {
```
Comment (kurnevsky) [inline src/toxcore/onion/packet/mod.rs]: Why is it handled separately?

File: src/toxcore/dht/codec.rs (hunk 2 above)
Comment (kpp) [reply src/toxcore/dht/codec.rs]: How?

Comment (kurnevsky) [reply src/toxcore/dht/codec.rs]: Create one unified `DecodeError::Deserialize` variant and place there `nom::Err` rather than placing `Incomplete` and `Error` separately.

Comment (kpp) [reply src/toxcore/dht/codec.rs]: Looks good

File: src/toxcore/dht/codec.rs (hunk 3 above)
Comment (kpp) [reply src/toxcore/dht/codec.rs]: @NamsooCho please fix

File: Cargo.toml (hunk 1 above)
Comment (NamsooCho) [reply Cargo.toml]: I am waiting release of 5.0.0.

File: src/toxcore/dht/codec.rs (hunk 3 above)
Comment (NamsooCho) [reply src/toxcore/dht/codec.rs]: done

Comment (NamsooCho) [reply src/toxcore/dht/codec.rs]: Maybe must change String to Vec<u8>.

Comment (kpp) [reply src/toxcore/dht/codec.rs]: Still there is `panic!` macro in your code. Please fix

File: src/toxcore/dht/daemon_state.rs (hunk 5 above)
Comment (kpp) [reply src/toxcore/dht/daemon_state.rs]: Why?

File: src/toxcore/dht/packet/errors.rs [hunk 8]
```diff
@@ -96,8 +91,8 @@ impl From<GetPayloadError> for IoError {
 
 #[cfg(test)]
 mod tests {
-    use super::*;
-
+//    use super::*;
```
Comment (kpp) [inline src/toxcore/dht/packet/errors.rs]: Why did you commented out this line?

File: src/toxcore/onion/packet/mod.rs (hunk 7 above)
Comment (kpp) [reply src/toxcore/onion/packet/mod.rs]: Why?

File: src/toxcore/state_format/old.rs [hunk 9]
```diff
@@ -346,14 +349,14 @@ pub struct FriendState {
     friend_status: FriendStatus,
     pk: PublicKey,
     /// Friend request message that is being sent to friend.
-    fr_msg: Vec<u8>,
+    pub fr_msg: Vec<u8>,
```
Comment (kpp) [inline src/toxcore/state_format/old.rs]: Why did you change from private to public?

File: src/toxcore/dht/daemon_state.rs (hunk 5 above)
Comment (NamsooCho) [reply src/toxcore/dht/daemon_state.rs]: I changed take! --> bytes::complete::take.
And now `Eof` is correct.
https://github.com/Geal/nom/pull/937/files#diff-5fbfee103ff92c86f212b5ae921ee6f4R373

File: src/toxcore/state_format/old.rs (hunk 9 above)
Comment (NamsooCho) [reply src/toxcore/state_format/old.rs]: done

File: src/toxcore/dht/codec.rs (hunk 3 above)
Comment (NamsooCho) [reply src/toxcore/dht/codec.rs]: https://github.com/tox-rs/tox/pull/374/files#diff-24a8beb3c46ed160ba513a45d8432efaR594

File: src/toxcore/dht/packet/cookie.rs [hunk 10]
```diff
@@ -235,18 +232,24 @@ mod tests {
             payload: invalid_payload_encoded
         };
         let decoded_payload = invalid_encrypted_cookie.get_payload(&symmetric_key);
-        assert!(decoded_payload.is_err());
-        assert_eq!(*decoded_payload.err().unwrap().kind(), GetPayloadErrorKind::Deserialize { error: ErrorKind::Eof, payload: invalid_encrypted_cookie.payload });
+        let error = decoded_payload.err().unwrap();
+        assert_eq!(*error.kind(), GetPayloadErrorKind::Deserialize {
+            error: Err::Error(("***************************************************".to_owned(), ErrorKind::Eof)),
```
Comment (kurnevsky) [inline src/toxcore/dht/packet/cookie.rs]: What do these stars mean?

File: src/toxcore/dht/codec.rs (hunk 2 above)
Comment (kurnevsky) [reply src/toxcore/dht/codec.rs]: Not fixed?

File: src/toxcore/dht/daemon_state.rs (hunk 4 above)
Comment (kurnevsky) [reply src/toxcore/dht/daemon_state.rs]: Not fixed?

File: src/toxcore/dht/packet/crypto_data.rs [hunk 11]
```diff
@@ -151,7 +148,9 @@ impl ToBytes for CryptoDataPayload {
 #[cfg(test)]
 mod tests {
     use super::*;
-    use nom::Needed;
+    use nom::{Err,
+        error::ErrorKind
+    };
```
Comment (kurnevsky) [inline src/toxcore/dht/packet/crypto_data.rs]: Something wrong with formatting.

File: src/toxcore/friend_connection/packet/share_relays.rs [hunk 12]
```diff
@@ -32,7 +32,7 @@ impl FromBytes for ShareRelays {
     named!(from_bytes<ShareRelays>, do_parse!(
         tag!(&[PACKET_ID_SHARE_RELAYS][..]) >>
         relays: many0!(PackedNode::from_tcp_bytes) >>
-        cond_reduce!(relays.len() <= MAX_SHARED_RELAYS, eof!()) >>
+        verify!(value!(relays.len()), |len| *len <= 3) >>
```
Comment (kurnevsky) [inline src/toxcore/friend_connection/packet/share_relays.rs]: Use `MAX_SHARED_RELAYS` instead of 3?

File: src/toxcore/onion/packet/onion_request_2.rs [hunk 13]
```diff
@@ -38,15 +38,15 @@ pub struct OnionRequest2 {
...
 impl FromBytes for OnionRequest2 {
     named!(from_bytes<OnionRequest2>, do_parse!(
-        verify!(rest_len, |len| len <= ONION_MAX_PACKET_SIZE) >>
+        verify!(rest_len, |len| *len <= ONION_MAX_PACKET_SIZE) >>
         tag!(&[0x82][..]) >>
         nonce: call!(Nonce::from_bytes) >>
         temporary_pk: call!(PublicKey::from_bytes) >>
         rest_len: rest_len >>
-        payload: cond_reduce!(
+        payload: map_opt!(cond!(
```
Comment (kurnevsky) [inline src/toxcore/onion/packet/onion_request_2.rs]: How about `rest_len: verify!(rest_len, |rest_len| rest_len >= ONION_RETURN_2_SIZE)` instead of `cond`+`map_opt`?

File: src/toxcore/dht/packet/cookie.rs (hunk 10 above)
Comment (NamsooCho) [reply src/toxcore/dht/packet/cookie.rs]: https://github.com/tox-rs/tox/pull/374/files#diff-2655c6396027ac3efcd54fe35a337272R32

https://github.com/tox-rs/tox/pull/374/files#diff-2655c6396027ac3efcd54fe35a337272R67

File: src/toxcore/onion/packet/onion_request_2.rs (hunk 13 above)
Comment (NamsooCho) [reply src/toxcore/onion/packet/onion_request_2.rs]: done

File: src/toxcore/binary_io.rs [hunk 14]
```diff
@@ -74,7 +74,7 @@ impl ToBytes for Ipv6Addr {
 
 /// Parser that returns the length of the remaining input.
 pub fn rest_len(input: &[u8]) -> IResult<&[u8], usize> {
-    IResult::Done(input, input.len())
+    Ok((input, input.len()))
```
Comment (kurnevsky) [inline src/toxcore/binary_io.rs]: There is rest_len in nom which we can use instead of this custom implementation.

File: src/toxcore/onion/packet/onion_announce_request.rs [hunk 15]
```diff
@@ -139,11 +136,11 @@ pub struct OnionAnnounceRequest {
 
 impl FromBytes for OnionAnnounceRequest {
     named!(from_bytes<OnionAnnounceRequest>, do_parse!(
-        rest_len: verify!(rest_len, |len| len <= ONION_MAX_PACKET_SIZE) >>
-        inner: cond_reduce!(
+        rest_len: verify!(rest_len, |len| *len <= ONION_MAX_PACKET_SIZE) >>
+        inner: map_opt!(cond!(
```
Comment (kurnevsky) [inline src/toxcore/onion/packet/onion_announce_request.rs]: The same should be done here as for `OnionRequest2`.

File: src/toxcore/onion/packet/onion_data_request.rs [hunk 16]
```diff
@@ -145,11 +140,11 @@ pub struct OnionDataRequest {
 
 impl FromBytes for OnionDataRequest {
     named!(from_bytes<OnionDataRequest>, do_parse!(
-        rest_len: verify!(rest_len, |len| len <= ONION_MAX_PACKET_SIZE) >>
-        inner: cond_reduce!(
+        rest_len: verify!(rest_len, |len| *len <= ONION_MAX_PACKET_SIZE) >>
+        inner: map_opt!(cond!(
```
Comment (kurnevsky) [inline src/toxcore/onion/packet/onion_data_request.rs]: The same should be done here as for `OnionRequest2`.

File: src/toxcore/onion/packet/onion_request_1.rs [hunk 17]
```diff
@@ -43,15 +43,15 @@ pub struct OnionRequest1 {
...
 impl FromBytes for OnionRequest1 {
     named!(from_bytes<OnionRequest1>, do_parse!(
-        verify!(rest_len, |len| len <= ONION_MAX_PACKET_SIZE) >>
+        verify!(rest_len, |len| *len <= ONION_MAX_PACKET_SIZE) >>
         tag!(&[0x81][..]) >>
         nonce: call!(Nonce::from_bytes) >>
         temporary_pk: call!(PublicKey::from_bytes) >>
         rest_len: rest_len >>
-        payload: cond_reduce!(
+        payload: map_opt!(cond!(
```
Comment (kurnevsky) [inline src/toxcore/onion/packet/onion_request_1.rs]: The same should be done here as for `OnionRequest2`.

File: src/toxcore/dht/packet/nodes_response.rs [hunk 18]
```diff
@@ -145,10 +142,10 @@ impl ToBytes for NodesResponsePayload {
 impl FromBytes for NodesResponsePayload {
     named!(from_bytes<NodesResponsePayload>, do_parse!(
         nodes_number: le_u8 >>
-        nodes: cond_reduce!(
+        nodes: map_opt!(cond!(
```
Comment (kurnevsky) [inline src/toxcore/dht/packet/nodes_response.rs]: The same should be done here as for `OnionRequest2`.

File: src/toxcore/dht/packet/cookie.rs (hunk 10 above)
Comment (kurnevsky) [reply src/toxcore/dht/packet/cookie.rs]: Then it's wrong to convert it to string, it's an unparsed input which is not a string. It should be bytes array.

File: src/toxcore/dht/packet/nodes_response.rs (hunk 18 above)
Comment (NamsooCho) [reply src/toxcore/dht/packet/nodes_response.rs]: done

File: src/toxcore/dht/daemon_state.rs [hunk 19]
```diff
@@ -147,14 +142,14 @@ mod tests {
         let serialized_vec = DaemonState::serialize_old(&alice);
         let serialized_len = serialized_vec.len();
         let res = DaemonState::deserialize_old(&alice, &serialized_vec[..serialized_len - 1]).wait();
-        assert!(res.is_err());
-        assert_eq!(*res.err().unwrap().kind(), DeserializeErrorKind::IncompleteData { needed: Needed::Size(55), data: serialized_vec[..serialized_len - 1].to_vec() });
+        res.err().unwrap();
```
Comment (kurnevsky) [inline src/toxcore/dht/daemon_state.rs]: Check the error?

File: src/toxcore/dht/daemon_state.rs [hunk 20]
```diff
@@ -147,14 +142,14 @@ mod tests {
...
 
         // test with serialized data corrupted
         let serialized_vec = [42; 10];
         let res = DaemonState::deserialize_old(&alice, &serialized_vec).wait();
-        assert!(res.is_err());
-        assert_eq!(*res.err().unwrap().kind(), DeserializeErrorKind::Deserialize { error: NomErrorKind::Tag, data: serialized_vec.to_vec() });
+        let error = res.err().unwrap();
+        assert_eq!(*error.kind(), DeserializeErrorKind::Deserialize { error: Err::Error((
+            vec![42, 42, 42, 42, 42, 42, 42, 42, 42, 42], NomErrorKind::Tag)), data: serialized_vec.to_vec() });
```
Comment (kurnevsky) [inline src/toxcore/dht/daemon_state.rs]: vec![42; 10]

File: src/toxcore/dht/packet/cookie.rs [hunk 21]
```diff
@@ -235,18 +232,27 @@ mod tests {
             payload: invalid_payload_encoded
         };
         let decoded_payload = invalid_encrypted_cookie.get_payload(&symmetric_key);
-        assert!(decoded_payload.is_err());
-        assert_eq!(*decoded_payload.err().unwrap().kind(), GetPayloadErrorKind::Deserialize { error: ErrorKind::Eof, payload: invalid_encrypted_cookie.payload });
+        let error = decoded_payload.err().unwrap();
+        assert_eq!(*error.kind(), GetPayloadErrorKind::Deserialize {
+            error: Err::Error((vec![42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42,
```
Comment (kurnevsky) [inline src/toxcore/dht/packet/cookie.rs]: vec![42; n]

File: src/toxcore/dht/packet/cookie_response.rs [hunk 22]
```diff
@@ -214,8 +211,11 @@ mod tests {
             payload: invalid_payload_encoded
         };
         let decoded_payload = invalid_packet.get_payload(&shared_secret);
-        assert!(decoded_payload.is_err());
-        assert_eq!(*decoded_payload.err().unwrap().kind(), GetPayloadErrorKind::Deserialize { error: ErrorKind::Eof, payload: invalid_packet.payload });
+        let error = decoded_payload.err().unwrap();
+        assert_eq!(*error.kind(), GetPayloadErrorKind::Deserialize {
+            error: Err::Error((vec![42, 42, 42], ErrorKind::Eof)),
```
Comment (kurnevsky) [inline src/toxcore/dht/packet/cookie_response.rs]: vec![42; 3]

File: src/toxcore/onion/packet/onion_announce_response.rs [hunk 23]
```diff
@@ -142,7 +139,7 @@ impl FromBytes for OnionAnnounceResponsePayload {
         announce_status: call!(AnnounceStatus::from_bytes) >>
         ping_id_or_pk: call!(sha256::Digest::from_bytes) >>
         nodes: many0!(PackedNode::from_bytes) >>
-        cond_reduce!(nodes.len() <= 4, eof!()) >>
+        cond!(nodes.len() <= 4, eof!()) >>
```
Comment (kurnevsky) [inline src/toxcore/onion/packet/onion_announce_response.rs]: It seems we need `verify` here? `cond` just returns an option.

File: src/toxcore/tcp/codec.rs [hunk 24]
```diff
@@ -109,13 +111,10 @@ impl Decoder for Codec {
...
-            IResult::Incomplete(_) => {
-                return Ok(None)
-            },
-            IResult::Error(error) => {
-                return Err(DecodeError::DeserializeEncryptedError { error, buf: buf.to_vec() })
-            },
-            IResult::Done(i, encrypted_packet) => {
+            Err(Err::Incomplete(_)) => return Ok(None),
+            Err(Err::Error(_)) => return Ok(None),
+            Err(Err::Failure(_)) => return Ok(None),
```
Comment (kurnevsky) [inline src/toxcore/tcp/codec.rs]: Why ok?

File: src/toxcore/tcp/codec.rs [hunk 25]
```diff
@@ -345,7 +347,7 @@ mod tests {
         let mut alice_codec = Codec::new(alice_channel, stats);
 
         // not enought bytes to decode EncryptedPacket
-        assert!(alice_codec.decode(&mut buf).is_err());
+        assert_eq!(alice_codec.decode(&mut buf).unwrap(), None);
```
Comment (kurnevsky) [inline src/toxcore/tcp/codec.rs]: Why?

File: src/toxcore/tcp/codec.rs (hunk 24 above)
Comment (NamsooCho) [reply src/toxcore/tcp/codec.rs]: If it is not ok then this happens

`toxcore::tcp::client::client::tests::spawn ... test toxcore::tcp::client::client::tests::spawn has been running for over 60 seconds`

But I can't resolve it why test hangs.

File: Cargo.toml (hunk 1 above)
Comment (kpp) [reply Cargo.toml]: There is an official release of nom5

**Comment by kpp** [comment]
Please fix CI builds

**Comment by NamsooCho** [comment]
https://github.com/Geal/nom/issues/971

File: src/toxcore/dht/codec.rs [hunk 26]
```diff
@@ -347,8 +343,8 @@ mod tests {
 
         let res = codec.decode(&mut buf);
         // not enought bytes to decode EncryptedPacket
-        assert!(res.is_err());
-        assert_eq!(*res.err().unwrap().kind(), DecodeErrorKind::Deserialize { error: ErrorKind::Alt, packet: vec![0xff] });
+        let error = res.err().unwrap();
+        assert_eq!(*error.kind(), DecodeErrorKind::Deserialize { error: Err::Error((vec![255], ErrorKind::Alt)) , packet: vec![0xff] });
```
Comment (kpp) [inline src/toxcore/dht/codec.rs]: `vec![255]` -> `vec![0xff]`

File: src/toxcore/dht/packet/cookie.rs [hunk 27]
```diff
@@ -235,18 +232,24 @@ mod tests {
...
-        assert!(decoded_payload.is_err());
-        assert_eq!(*decoded_payload.err().unwrap().kind(), GetPayloadErrorKind::Deserialize { error: ErrorKind::Eof, payload: invalid_encrypted_cookie.payload });
+        let error = decoded_payload.err().unwrap();
+        assert_eq!(*error.kind(), GetPayloadErrorKind::Deserialize {
+            error: Err::Error((vec![42; 51], ErrorKind::Eof)),
+            payload: invalid_payload.to_vec()
+        });
         // Try short incomplete array
-        let invalid_payload = [];
+        let invalid_payload = [42, 42];
```
Comment (kpp) [inline src/toxcore/dht/packet/cookie.rs]: Why did you change [] to [42, 42]?

File: src/toxcore/dht/packet/errors.rs [hunk 28]
```diff
@@ -52,24 +55,16 @@ impl fmt::Display for GetPayloadError {
 }
 
 /// The specific kind of error that can occur.
-#[derive(Clone, Debug, Eq, PartialEq, Fail)]
```
Comment (kpp) [inline src/toxcore/dht/packet/errors.rs]: Why not Eq? You should create an issue in nom to make nom::Err Eq

File: src/toxcore/messenger/packet/msi.rs [hunk 29]
```diff
@@ -206,8 +209,9 @@ pub struct Msi {
     capabilities: Capabilities,
 }
 
-const NOM_CUSTOM_ERR_REQUEST_SUBPACKET_OMITTED: u32 = 1;
-const NOM_CUSTOM_ERR_CAPABILITIES_SUBPACKET_OMITTED: u32 = 2;
+//FIXME : use custom ErrorKind
```
Comment (kpp) [inline src/toxcore/messenger/packet/msi.rs]: Do we need these comments?

File: src/toxcore/onion/packet/onion_announce_request.rs [hunk 30]
```diff
@@ -139,11 +137,8 @@ pub struct OnionAnnounceRequest {
 
 impl FromBytes for OnionAnnounceRequest {
     named!(from_bytes<OnionAnnounceRequest>, do_parse!(
-        rest_len: verify!(rest_len, |len| len <= ONION_MAX_PACKET_SIZE) >>
-        inner: cond_reduce!(
-            rest_len >= ONION_RETURN_3_SIZE,
-            flat_map!(take!(rest_len - ONION_RETURN_3_SIZE), InnerOnionAnnounceRequest::from_bytes)
-        ) >>
+        rest_len: verify!(rest_len, |len| *len <= ONION_MAX_PACKET_SIZE && *len >= ONION_RETURN_3_SIZE) >>
```
Comment (kpp) [inline src/toxcore/onion/packet/onion_announce_request.rs]: @kurnevsky is it OK?

File: src/toxcore/onion/packet/onion_announce_response.rs [hunk 31]
```diff
@@ -142,7 +139,8 @@ impl FromBytes for OnionAnnounceResponsePayload {
         announce_status: call!(AnnounceStatus::from_bytes) >>
         ping_id_or_pk: call!(sha256::Digest::from_bytes) >>
         nodes: many0!(PackedNode::from_bytes) >>
-        cond_reduce!(nodes.len() <= 4, eof!()) >>
+        _len: verify!(value!(nodes.len()), |len| *len <= 4 as usize) >>
```
Comment (kpp) [inline src/toxcore/onion/packet/onion_announce_response.rs]: You don't need to create an unused variable `_len`, you may use `verify!` without `_len: ` IIRC.

File: src/toxcore/onion/packet/onion_data_request.rs [hunk 32]
```diff
@@ -145,11 +142,8 @@ pub struct OnionDataRequest {
 
 impl FromBytes for OnionDataRequest {
     named!(from_bytes<OnionDataRequest>, do_parse!(
-        rest_len: verify!(rest_len, |len| len <= ONION_MAX_PACKET_SIZE) >>
-        inner: cond_reduce!(
-            rest_len >= ONION_RETURN_3_SIZE,
-            flat_map!(take!(rest_len - ONION_RETURN_3_SIZE), InnerOnionDataRequest::from_bytes)
-        ) >>
+        rest_len: verify!(rest_len, |len| *len <= ONION_MAX_PACKET_SIZE && *len >= ONION_RETURN_3_SIZE) >>
```
Comment (kpp) [inline src/toxcore/onion/packet/onion_data_request.rs]: @kurnevsky is it OK?

File: src/toxcore/tcp/codec.rs [hunk 33]
```diff
@@ -109,13 +111,10 @@ impl Decoder for Codec {
...
         let (consumed, encrypted_packet) = match EncryptedPacket::from_bytes(buf) {
-            IResult::Incomplete(_) => {
-                return Ok(None)
-            },
-            IResult::Error(error) => {
-                return Err(DecodeError::DeserializeEncryptedError { error, buf: buf.to_vec() })
-            },
-            IResult::Done(i, encrypted_packet) => {
+            Err(Err::Incomplete(_)) => return Ok(None),
+            Err(Err::Error(_)) => return Ok(None),
```
Comment (kpp) [inline src/toxcore/tcp/codec.rs]: I am not quite sure we have to return `Ok(None)` if we got nom::Err::Error. If we return `Ok(None)` we tell tokio::Codec to leave the buffer untouched and try to fill it with a new portion of bytes. So in some scenario we will get nom::Err::Error and we will return `Ok(None)` so the buffer will grow and grow and we will got a DOS.

File: src/toxcore/dht/codec.rs (hunk 26 above)
Comment (kpp) [reply src/toxcore/dht/codec.rs]: @NamsooCho please fix

File: src/toxcore/tcp/codec.rs (hunk 33 above)
Comment (Geal) [reply src/toxcore/tcp/codec.rs]: sorry for the delay. @kpp: `Err::Error` and `Err::Failure` have the same effect everywhere, except for combinators like `alt`. If one of the child parsers returns `Err:Error`, `alt` tries the next one, but if it returned `Err:Failure`, the error is returned right away, and we do not try any other branch. `Failure` indicates that we were in the right branch, but something went wrong. The `cut` combinator can be used to wrap another parser and transform its `Err::Error` results to `Err::Failure`

File: src/toxcore/onion/packet/onion_announce_request.rs (hunk 30 above)
Comment (kpp) [reply src/toxcore/onion/packet/onion_announce_request.rs]: @kurnevsky ping

**Comment by kpp** [comment]
https://github.com/tox-rs/tox/pull/418
